pkg/resources/management.cattle.io/v3/authconfig/validator.go

```diff
@@ -0,0 +1,272 @@
+	}
+	if config.GroupSearchFilter != "" {
+		if _, fieldErr := ldapv3.CompileFilter(config.GroupSearchFilter); fieldErr != nil {
+			err = errors.Join(err, field.Forbidden(field.NewPath("groupSearchFilter"), fmt.Sprintf("%s", fieldErr)))
+		}
+	}
+
+	return err
+}
+
+func validateActiveDirectoryConfig(request *admission.Request) error {
+	var (
+		err    error
+		config v3.ActiveDirectoryConfig
+	)
+
+	if err = json.Unmarshal(request.Object.Raw, &config); err != nil {
+		return fmt.Errorf("failed to get %T from request: %w", config, err)
+	}
+
+	if len(config.Servers) < 1 {
+		err = errors.Join(err, field.Forbidden(field.NewPath("servers"), "at least one server is required"))
+	}
+	if config.TLS && config.Certificate == "" {
+		err = errors.Join(err, field.Forbidden(field.NewPath("certificate"), "certificate is required"))
+	}
+
+	if config.UserSearchAttribute != "" {
+		for _, attr := range strings.Split(config.UserSearchAttribute, "|") {
+			if !isValidAttr(attr) {
+				err = errors.Join(err, field.Forbidden(field.NewPath("userSearchAttribute"), "invalid value"))
+			}
+		}
+	}
+	if config.UserLoginAttribute != "" && !isValidAttr(config.UserLoginAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("userLoginAttribute"), "invalid value"))
+	}
+	if config.UserObjectClass != "" && !isValidAttr(config.UserObjectClass) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("userObjectClass"), "invalid value"))
+	}
+	if config.UserNameAttribute != "" && !isValidAttr(config.UserNameAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("userNameAttribute"), "invalid value"))
+	}
+	if config.UserEnabledAttribute != "" && !isValidAttr(config.UserEnabledAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("userEnabledAttribute"), "invalid value"))
+	}
+	if config.GroupSearchAttribute != "" && !isValidAttr(config.GroupSearchAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupSearchAttribute"), "invalid value"))
+	}
+	if config.GroupObjectClass != "" && !isValidAttr(config.GroupObjectClass) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupObjectClass"), "invalid value"))
+	}
+	if config.GroupNameAttribute != "" && !isValidAttr(config.GroupNameAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupNameAttribute"), "invalid value"))
+	}
+	if config.GroupDNAttribute != "" && !isValidAttr(config.GroupDNAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupDNAttribute"), "invalid value"))
+	}
+	if config.GroupMemberUserAttribute != "" && !isValidAttr(config.GroupMemberUserAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupMemberUserAttribute"), "invalid value"))
+	}
+	if config.GroupMemberMappingAttribute != "" && !isValidAttr(config.GroupMemberMappingAttribute) {
+		err = errors.Join(err, field.Forbidden(field.NewPath("groupMemberMappingAttribute"), "invalid value"))
+	}
+
+	if config.UserLoginFilter != "" {
```

Comment:
I wonder if these shouldn't be moved up to where we check the values so that the errors come out together when logged?

Reply:
I'm not sure I understand what you mean. All of these are independent fields e.g. LoginUserAttribute has nothing to do with UserLoginFilter.

Reply:
To clarify, you're using `errors.Join()` which under the covers returns an `errors.joinError` the `.Error()` method basically appends the `.Error()` of each error with `\n`.

It _might_ be nice if the two possible `UserSearchFilter` errors came out on consecutive lines?

Reply:
Right, this is exactly how it's ordered right now. If you have validation errors, at the top you'll see all errors regarding different attributes, while at the bottom all filter errors.

>  if the two possible UserSearchFilter errors came out on consecutive lines?

There can only be one possible error related to UserSearchFilter (or any other field for that matter), the one that says that it's invalid if the expression it contains can't be successfully compiled.
The only exception is `UserSearchAttribute` which contains a delimited list of attribute names but all possible error messages for this field are grouped together(going one after another) L207.